### PR TITLE
[FLINK-3714] Add Support for "Allowed Lateness"

### PR DESCRIPTION
**Pull request by kl0u**
Allows the user to specify an "allowed lateness" for  elements, before they are dropped as late arrivals.
In addition, it defines a "clean up" time for the window state, which is the end of the window plus the 
aforementioned allowed lateness. When the watermark or the processing time (depending on the widow
assigner) passes this mark, then the window state is purged/cleaned.

The design of this PR was initially discussed also with @aljoscha

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/WindowedStream.java [hunk 1]
```diff
@@ -121,6 +124,23 @@ public WindowedStream(KeyedStream<T, K> input,
 	}
 
 	/**
+	 * Sets the allowed lateness. If the {@link WindowAssigner} used
+	 * is in processing time, then the allowed lateness is set to 0.
+	 */
+	@PublicEvolving
+	public WindowedStream<T, K, W> setAllowedLateness(Time lateness) {
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/WindowedStream.java]: This should follow the same pattern as the other methods (`trigger()`, `evictor()`...), i.e. `allowedLateness(Time)`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/WindowedStream.java [hunk 2]
```diff
@@ -121,6 +124,23 @@ public WindowedStream(KeyedStream<T, K> input,
...
+		long millis = lateness.toMilliseconds();
+		if (allowedLateness < 0) {
+			throw new IllegalArgumentException("The allowed lateness cannot be negative.");
+		} else if (allowedLateness != 0 && !windowAssigner.isEventTime()) {
+			this.allowedLateness = 0;
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/WindowedStream.java]: We probably should log a warning here and not silently set to zero.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/EventTimeTrigger.java [hunk 3]
```diff
@@ -36,12 +36,17 @@ private EventTimeTrigger() {}
 	@Override
 	public TriggerResult onElement(Object element, long timestamp, TimeWindow window, TriggerContext ctx) throws Exception {
 		ctx.registerEventTimeTimer(window.maxTimestamp());
-		return TriggerResult.CONTINUE;
+
+		return (timestamp < ctx.getCurrentWatermark()) ?
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/EventTimeTrigger.java]: This should be `window.maxTimestamp() <= ctx.getCurrentWatermark()`. If the tests pass with this, then they also need to be adapted to fail with the current code and then work with the new code.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/WindowedStream.java (hunk 2 above)
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/WindowedStream.java]: I agree but here we do not have a logger here yet.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java [hunk 4]
```diff
@@ -129,6 +130,12 @@
 	 */
 	protected final TypeSerializer<W> windowSerializer;
 
+	/**
+	 * The allowed lateness for elements.
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java]: This needs to be updated. The allowed lateness is used both for dropping of late elements (based on the window into which they fall) and garbage collection of window state). The default is not set in this operator but in `WindowedStream`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java [hunk 5]
```diff
@@ -264,7 +274,9 @@ public void dispose() {
 	@Override
 	@SuppressWarnings("unchecked")
 	public void processElement(StreamRecord<IN> element) throws Exception {
-		Collection<W> elementWindows = windowAssigner.assignWindows(element.getValue(), element.getTimestamp());
+
+		Collection<W> elementWindows = windowAssigner.
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java]: Here I'm just being nitpicky but the code looks strange if broken into lines like this. To me
```
Collection<W> elementWindows =
    windowAssigner.assignWindows(element.getValue(), element.getTimestamp());
```
or
```
Collection<W> elementWindows = windowAssigner.assignWindows(
    element.getValue(),
    element.getTimestamp()); 
```

seem more elegant.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/WindowedStream.java (hunk 2 above)
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/WindowedStream.java]: Then we can add a `LOG`, or maybe we should even throw an exception. What do you think?

Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/WindowedStream.java]: LOG is more natural. There is no reason to stop the job for the user (as it is not an error). We should just tell him that 
it does not make that much sense, so that he knows how to interpret the results he gets. Throwing an exception is somehow more drastic.

Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/WindowedStream.java]: Yeah, but it would throw right when specifying the job. Not when running or anything. LOG is also fine, though. 😃

**Comment by aljoscha** [comment]
I had some inline comments.

One other thing I would like to see changed is moving the logic that checks whether to do cleanup from `cleanup()` to the call-site. So in `processElement()` it would be:
```
if (combinedTriggerResult.isPurge) {
  cleanup(...);
} else {
  registerCleanupTimer(...);
}
```
The reason for this is that the check in `cleanup()` for whether a timer is a cleanup timer does not take into account the event-time/processing-time split: if a processing-time timer fires on exactly the cleanup time of an event-time window this would also trigger cleanup.  The code in `processTriggersFor` (which, by the way can be moved back into `processWatermark()` now that you refactored this special event-time trigger hack) would have this check:
```
if (assigner.isEventTime() && isCleanupTimer(...))
```
while the code in `trigger()` would be
```
if (!assigner.isEventTime() && isCleanupTimer(...))
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java [hunk 6]
```diff
@@ -551,19 +657,28 @@ public void registerProcessingTimeTimer(long time) {
 			}
 		}
 
+		void registerCleanupTimer(long time) {
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java]: The logic for these two (`register/deleteCleanupTimer`) should be moved directly into `WindowOperator`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java [hunk 7]
```diff
@@ -131,8 +131,10 @@
 	protected final TypeSerializer<W> windowSerializer;
 
 	/**
-	 * The allowed lateness for elements.
-	 * By default this is set to {@code 0}.
+	 * The allowed lateness for elements. This is used for:
+	 * <li> Deciding if an element should be dropped from a window due to lateness. </li>
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java]: I think you need an `<ul>` around this whole list. And the closing `</li>` is not used in Javadoc, I think.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/WindowedStream.java [hunk 8]
```diff
@@ -124,16 +124,17 @@ public WindowedStream(KeyedStream<T, K> input,
 	}
 
 	/**
-	 * Sets the allowed lateness. If the {@link WindowAssigner} used
+	 * Sets the allowed lateness. By default this is 0. If the {@link WindowAssigner} used
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/WindowedStream.java]: The javadoc is out of date with the new exception-throwing behavior. For the text of the exception I would prefer something like "Setting the allowed lateness is only valid for event-time windows.". Precise and simple. 😃

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java [hunk 9]
```diff
@@ -450,17 +436,140 @@ public final void trigger(long time) throws Exception {
...
+			mergingWindows = new MergingWindowSet<>((MergingWindowAssigner<? super IN, W>) windowAssigner, mergeState);
+			mergeState.clear();
+
+			mergingWindowsByKey.put((K) getStateBackend().getCurrentKey(), mergingWindows);
+		}
+		return mergingWindows;
+	}
+
+	/**
+	 * This method decides if a window is currently active, or not, based on the current
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java]: I would prefer this to start with "Decides if a window is late nor not, based on ...". It is not clear what `active` means here and `late` describes pretty well what the method does.

**Comment by aljoscha** [comment]
The changes look very good now! The refactoring of the cleanup and firing makes it very easy to see what's going on. I just had some inline comments about outdated Javadoc and exception strings.

I still have to read the tests in depth, I'm doing that now.

**Comment by aljoscha** [comment]
Can you please also add a short javadoc on `EventTimeTriggerAccum`, i.e. what it does and why we need a special trigger for the tests.

**Comment by aljoscha** [comment]
For the tests, I think you need to compare the expected results to the actual results the same way the existing tests do it, via exact comparison. As you put it now, it would allow that all actual results are the same value because the tests don't verify that we see all expected values, just that the size of the result list matches.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java [hunk 10]
```diff
@@ -132,9 +132,9 @@
 
 	/**
 	 * The allowed lateness for elements. This is used for:
-	 * <li> Deciding if an element should be dropped from a window due to lateness. </li>
-	 * <li> Clearing the state of a window if the system time passes
-	 * the {@code window.maxTimestamp + allowedLateness} landmark. </li>
+	 * <ul> Deciding if an element should be dropped from a window due to lateness.</ul>
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java]: Ah, I meant something like this:
```
<ul>
  <li> item 1
  <li> item2
</ul>

the javadoc syntax is almost similar to HTML, just the closing `</li>` tags are missing.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorTest.java [hunk 11]
```diff
@@ -917,6 +929,465 @@ public void testRestoreAndSnapshotAreInSync() throws Exception {
...
+
+		expected.add(new Watermark(initialTime + 6000));
+		expected.add(new Watermark(initialTime + 7000));
+
+		TestHarnessUtil.assertOutputEqualsSorted("Output was not correct.", expected, testHarness.getOutput(), new Tuple2ResultSortComparator());
+		testHarness.close();
+	}
+
+	@Test
+	public void testDropDueToLatenessTumbling() throws Exception {
```
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorTest.java]: No element is dropped in this test. If I'm not mistaken.

**Comment by aljoscha** [comment]
Could you please update the tests to add elements to `expectedOutput` in line with the `processElement` and `processWatermark` calls on the test harness. In the existing tests I did this and it makes it easy to reason about what output should be produced in sequence. If the expected elements/watermark are added in after the other calls you have to keep jumping back and forth when reading the test. Also, could you replace the `el`, `el2` and so on variables with inline tuple creation. For the same reason of not having to jump back and forth when reading the tests?

**Comment by aljoscha** [comment]
Some tests, such as `testLateness()` still fill the expected output in one block at the end of the test and not inline with the `processElement()`/`processWatermark()` calls. Some, such as `testDropDueToLatenessSliding` still have the variables `el1`, `el2` and so on that make it hard to parse what is going on.

Please also fix those tests. The session tests look very good now!

**Comment by kl0u** [comment]
Thanks for merging it @aljoscha ! I am closing this PR and the issue related to it.
